microscope/testsuite/libs.py: new module to patch ctypes.CDLL with mock C libs

carandraug opened the pull request:
At the moment, very simple. It provides a fake CDLL class to patch ctypes.CDLL. Importing the module does not actually do anything, one then needs to actually do the patching.  Tested in Python 2 and 3:

```
$ python3
Python 3.4.2 (default, Oct  8 2014, 10:45:20) 
[GCC 4.9.1] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import microscope.testsuite.libs
>>> import ctypes
>>> ctypes.CDLL = microscope.testsuite.libs.CDLL
>>> import microscope.cameras.SDK3
>>> import microscope.cameras.pvcam
>>> import microscope.mirror.alpao
>>> import microscope.mirror.bmc
>>> import microscope._wrappers.BMC
>>> import microscope._wrappers.asdk
>>> 

$ python
Python 2.7.9 (default, Jun 29 2016, 13:08:31) 
[GCC 4.9.2] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import microscope.testsuite.libs
>>> import ctypes
>>> ctypes.CDLL = microscope.testsuite.libs.CDLL
>>> import microscope.cameras.SDK3
>>> import microscope.cameras.pvcam
>>> import microscope.mirror.alpao
>>> import microscope.mirror.bmc
>>> import microscope._wrappers.BMC
>>> import microscope._wrappers.asdk
```

It's not yet making use of unittest.mock (there is a mock package which backports it for python2) but that can be done after. For know, this is enough to allow us to import all modules without the C libraries installed and build the documentation.  The `unittest.mock` module provides us with a patch decorator and context that we can make further use in the test units.

mickp commented:
I think this is great, and doesn't need to do more - testing beyond import needs hardware, really. Knowing we can still import after making changes elsewhere (i.e. our changes or dependency changes haven't broken anything unexpected) is probably the best we can do without hardware, or a thorough emulation of the hardware.

carandraug commented:
I made one more modification so that by default functions in the mocked library raise an exception. This will help us identify if they are being called somewhere.  I have also added an implementation for `AT_InitialiseUtilityLibrary` which is being called during import of `microscope.cameras.SDK3`.

Also added some more docs for doctest:

    python -m doctest  -v microscope/testsuite/libs.py

Would still be nice for in the future to have a CDLL class that would intercept a subset of libraries instead of all but I will leave for future work.

carandraug commented:
Rebased and squashed 00ee88906d
